drivers/timer/nrf_grtc_timer.c

```diff
@@ -0,0 +1,571 @@
+		       z_nrf_grtc_timer_compare_handler_t handler, void *user_data)
+{
+	bool key = compare_int_lock(chan);
+	int ret = compare_set_nolocks(chan, target_time, handler, user_data);
+
+	compare_int_unlock(chan, key);
+
+	return ret;
+}
+
+int z_nrf_grtc_timer_set(int32_t chan, uint64_t target_time,
+			 z_nrf_grtc_timer_compare_handler_t handler, void *user_data)
+{
+	IS_CHANNEL_ALLOWED_ASSERT(chan);
+
+	return compare_set(chan, target_time, (nrfx_grtc_cc_handler_t)handler, user_data);
+}
+
+void z_nrf_grtc_timer_abort(int32_t chan)
+{
+	IS_CHANNEL_ALLOWED_ASSERT(chan);
+
+	bool key = compare_int_lock(chan);
+	(void)nrfx_grtc_syscounter_cc_disable(chan);
+	compare_int_unlock(chan, key);
+}
+
+uint64_t z_nrf_grtc_timer_get_ticks(k_timeout_t t)
+{
+	uint64_t curr_time;
+	int64_t curr_tick;
+	int64_t result;
+	int64_t abs_ticks;
+
+	curr_time = counter();
+	curr_tick = sys_clock_tick_get();
+
+	abs_ticks = Z_TICK_ABS(t.ticks);
+	if (abs_ticks < 0) {
+		/* relative timeout */
+		return (t.ticks > (int64_t)COUNTER_SPAN) ? -EINVAL : (curr_time + t.ticks);
+	}
+
+	/* absolute timeout */
+	result = abs_ticks - curr_tick;
+
+	if (result > (int64_t)COUNTER_SPAN) {
+		return -EINVAL;
+	}
+
+	return curr_time + result;
+}
+
+int z_nrf_grtc_timer_capture_prepare(int32_t chan)
+{
+	nrfx_grtc_channel_t user_channel_data = {
+		.handler = NULL,
+		.p_context = NULL,
+		.channel = chan,
+	};
+	nrfx_err_t result;
+
+	IS_CHANNEL_ALLOWED_ASSERT(chan);
+
+	/* Set the CC value to mark channel as not triggered and also to enable it
+	 * (makes CCEN=1). COUNTER_SPAN is used so as not to fire an event unnecessarily
+	 * - it can be assumed that such a large value will never be reached.
+	 */
+	result = nrfx_grtc_syscounter_cc_absolute_set(&user_channel_data, COUNTER_SPAN, false);
+
+	if (result != NRFX_SUCCESS) {
+		return -EPERM;
+	}
+
+	return 0;
+}
+
+int z_nrf_grtc_timer_capture_read(int32_t chan, uint64_t *captured_time)
+{
+	/* TODO: The implementation should probably go to nrfx_grtc and this
+	 *       should be just a wrapper for some nrfx_grtc_syscounter_capture_read.
+	 */
+
+	uint64_t capt_time;
+
+	IS_CHANNEL_ALLOWED_ASSERT(chan);
+
+	/* TODO: Use `nrfy_grtc_sys_counter_enable_check` when available (NRFX-2480) */
+	if (NRF_GRTC->CC[chan].CCEN == GRTC_CC_CCEN_ACTIVE_Enable) {
+		/* If the channel is enabled (.CCEN), it means that there was no capture
+		 * triggering event.
+		 */
+		return -EBUSY;
+	}
+
+	capt_time = nrfy_grtc_sys_counter_cc_get(NRF_GRTC, chan);
+
+	__ASSERT_NO_MSG(capt_time < COUNTER_SPAN);
+
+	*captured_time = capt_time;
+
+	return 0;
+}
+
+#if defined(CONFIG_NRF_GRTC_SLEEP_ALLOWED)
+int z_nrf_grtc_wakeup_prepare(uint64_t wake_time_us)
+{
+	nrfx_err_t err_code;
+	static uint8_t systemoff_channel;
+	uint64_t now = counter();
+	/* Minimum time that ensures valid execution of system-off procedure. */
+	uint32_t minimum_latency_us =
+		nrfy_grtc_waketime_get(NRF_GRTC) + nrfy_grtc_timeout_get(NRF_GRTC) + 1000;
```

Comment:
why 1000?

Reply:
CC @adamkondraciuk

Reply:
It's the lowest value when you can be sure, that the wakeup event will not fire too early (of course plus TIMEOUT and WAKETIME register values) - I mean during preparing to sleep. Maybe it could be defined here: https://github.com/zephyrproject-rtos/zephyr/blob/a7a57e2a916bd7df9c4f84ce902f3082060295f6/drivers/timer/Kconfig.nrf_grtc
as e.g. `config NRF_GRTC_SLEEP_MINIMUM_LATENCY` or something like that?

Reply:
or even in this file but hidden behind some name and comment